Add `core::mem::DropGuard`

yoshuawuyts opened the pull request:
_Note: The functionality added in this PR is now being tracked in https://github.com/rust-lang/rust/issues/144426._

## 1.0 Summary

This PR introduces a new type `core::mem::DropGuard` which wraps a value and runs a closure when the value is dropped.

```rust
use core::mem::DropGuard;

// Create a new guard around a string that will
// print its value when dropped.
let s = String::from("Chashu likes tuna");
let mut s = DropGuard::new(s, |s| println!("{s}"));

// Modify the string contained in the guard.
s.push_str("!!!");

// The guard will be dropped here, printing:
// "Chashu likes tuna!!!"
```

## 2.0 Motivation

A number of programming languages include constructs like `try..finally` or `defer` to run code as the last piece of a particular sequence, regardless of whether an error occurred. This is typically used to clean up resources, like closing files, freeing memory, or unlocking resources. In Rust we use the `Drop` trait instead, allowing us to [never having to manually close sockets](https://blog.skylight.io/rust-means-never-having-to-close-a-socket/).

While `Drop` (and RAII in general) has been working incredibly well for Rust in general, sometimes it can be a little verbose to setup. In particular when upholding invariants are local to functions, having a quick inline way to setup an `impl Drop` can be incredibly convenient. We can see this in use in the Rust stdlib, which has a number of private `DropGuard` impls used internally:

- [library/alloc/src/vec/drain.rs](https://github.com/rust-lang/rust/blob/9982d6462bedf1e793f7b2dbd655a4e57cdf67d4/library/alloc/src/vec/drain.rs#L177)
- [library/alloc/src/boxed/thin.rs](https://github.com/rust-lang/rust/blob/9982d6462bedf1e793f7b2dbd655a4e57cdf67d4/library/alloc/src/boxed/thin.rs#L362)
- [library/alloc/src/slice.rs](https://github.com/rust-lang/rust/blob/9982d6462bedf1e793f7b2dbd655a4e57cdf67d4/library/alloc/src/slice.rs#L413)
- [library/alloc/src/collections/linked_list.rs](https://github.com/rust-lang/rust/blob/9982d6462bedf1e793f7b2dbd655a4e57cdf67d4/library/alloc/src/collections/linked_list.rs#L1135)
- [library/alloc/src/collections/binary_heap/mod.rs](https://github.com/rust-lang/rust/blob/9982d6462bedf1e793f7b2dbd655a4e57cdf67d4/library/alloc/src/collections/binary_heap/mod.rs#L1816)
- [library/alloc/src/collections/btree/map.rs](https://github.com/rust-lang/rust/blob/9982d6462bedf1e793f7b2dbd655a4e57cdf67d4/library/alloc/src/collections/btree/map.rs#L1715)
- [library/alloc/src/collections/vec_deque/drain.rs](https://github.com/rust-lang/rust/blob/9982d6462bedf1e793f7b2dbd655a4e57cdf67d4/library/alloc/src/collections/vec_deque/drain.rs#L95)
- [library/alloc/src/vec/into_iter.rs](https://github.com/rust-lang/rust/blob/9982d6462bedf1e793f7b2dbd655a4e57cdf67d4/library/alloc/src/vec/into_iter.rs#L488)
- [library/std/src/os/windows/process.rs](https://github.com/rust-lang/rust/blob/9982d6462bedf1e793f7b2dbd655a4e57cdf67d4/library/std/src/os/windows/process.rs#L320)
- [tests/ui/process/win-proc-thread-attributes.rs](https://github.com/rust-lang/rust/blob/9982d6462bedf1e793f7b2dbd655a4e57cdf67d4/tests/ui/process/win-proc-thread-attributes.rs#L17)

## 3.0 Design

This PR implements what can be considered about the simplest possible design:

1. A single type `DropGuard` which takes both a generic type `T` and a closure `F`.
2. `Deref` + `DerefMut` impls to make it easy to work with the `T` in the guard.
3. An `impl Drop` on the guard which calls the closure `F` on drop.
4. An inherent `fn into_inner` which takes the type `T` out of the guard without calling the closure `F`.

Notably this design does not allow divergent behavior based on the type of drop that has occurred. The [`scopeguard` crate](https://docs.rs/scopeguard/latest/scopeguard/index.html) includes additional `on_success` and `on_onwind` variants which can be used to branch on unwind behavior instead. However [in a lot of cases](https://github.com/rust-lang/rust/issues/143612#issuecomment-3053928328) this doesn’t seem necessary, and using the arm/disarm pattern seems to provide much the same functionality:

```rust
let guard = DropGuard::new((), |s| ...);  // 1. Arm the guard
other_function();                         // 2. Perform operations
guard.into_inner();                       // 3. Disarm the guard
```

`DropGuard` combined with this pattern seems like it should cover the vast majority of use cases for quick, inline destructors. It certainly seems like it should cover all existing uses in the stdlib, as well as all existing uses in crates like [hashbrown](https://github.com/search?q=repo%3Arust-lang%2Fhashbrown%20guard&type=code).

## 4.0 Acknowledgements

This implementation is based on the [mini-scopeguard crate](https://github.com/yoshuawuyts/mini-scopeguard) which in turn is based on the [scopeguard  crate](https://docs.rs/scopeguard). The implementations only differ superficially; because of the nature of the problem there is only really one obvious way to structure the solution. And the scopeguard crate got that right!

## 5.0 Conclusion

This PR adds a new type `core::mem::DropGuard` to the stdlib which adds a small convenience helper to create inline destructors with. This would bring the majority of the functionality of the `scopeguard` crate into the stdlib, which is the [49th most downloaded crate](https://crates.io/crates?sort=downloads) on crates.io (387 million downloads).

Given the actual implementation of `DropGuard` is only around 60 lines, it seems to hit that sweet spot of low-complexity / high-impact that makes for a particularly efficient stdlib addition. Which is why I’m putting this forward for consideration; thanks!

yoshuawuyts committed:
Add `core::mem::DropGuard`

Fix CI for drop_guard

fix CI

fix all tidy lints

fix tidy link

add first batch of feedback from review

Add second batch of feedback from review

add third batch of feedback from review

fix failing test

Update library/core/src/mem/drop_guard.rs

Co-authored-by: Ruby Lazuli <general@patchmixolydic.com>

fix doctests

Implement changes from T-Libs-API review

And start tracking based on the tracking issue.

fix tidy lint

workingjubilee commented:
this appears to be revisiting https://github.com/rust-lang/libs-team/issues/622

yoshuawuyts commented:
> this appears to be revisiting [rust-lang/libs-team#622](https://github.com/rust-lang/libs-team/issues/622)

Oh you mean the naming? I guess I missed that issue since it's just a few days old. Most of the names come from the [mini-scopeguard](https://github.com/yoshuawuyts/mini-scopeguard) crate I put together in preparation for this PR earlier this year. To give some rationale for the names:

- `core::mem::DropGuard` groups nicely with `core::mem::drop` and `core::mem::ManuallyDrop`. It would be the third drop-related operation we're putting in `mem`, and it's worth making it sound both similar yet distinctly identifiable.
- `core::mem::DropGuard` functions very similarly to [all the other guards](https://doc.rust-lang.org/std/?search=guard) we have in the stdlib. I think of the "guard" suffix as "does interesting things on drop", which checks out.
- In the stdlib we've been using the name `DropGuard` for many years now, with at least [10 instances of it](https://github.com/search?q=repo%3Arust-lang%2Frust+dropguard&type=code&repo=&langOverride=&start_value=1). That seems like pretty good precedent, and a decent indicator that the name is memorable.
- The `ScopeGuard` crate uses the [`ScopeGuard::into_inner` method](https://docs.rs/scopeguard/latest/scopeguard/struct.ScopeGuard.html#method.into_inner) to take a value without running the destructors. Its signature is also nearly identical to all the other [`into_inner`](https://doc.rust-lang.org/std/?search=into_inner) methods we have.

This to me seemed like a reasonable starting point for a contribution. The part I'm least sure about is the `into_inner` method, since it does something interesting that might be worth calling attention to in a way that `into_inner` might not. Since this is only an initial impl, that seems like something to raise as an open question.

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16406778526/job/46354014333) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46354014333) [(plain)](https://github.com/rust-lang/rust/commit/dc5f4539af88266b2d984851bb403bc93fa2ed73/checks/46354014333/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
[TIMING] core::build_steps::tool::ToolBuild { build_compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu, tool: "tidy", path: "src/tools/tidy", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 8.579
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
Diff in /checkout/library/core/src/mem/drop_guard.rs:138:
 // tests copied from https://docs.rs/scopeguard/latest/src/scopeguard/lib.rs.html#1-595
 #[cfg(test)]
 mod tests {
-    use crate::cell::Cell;
-
     use super::*;
```

</details>

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16407905087/job/46356914910) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46356914910) [(plain)](https://github.com/rust-lang/rust/commit/709463719a1a3d05f831ccf5c52f6711b3b7b260/checks/46356914910/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Checking tidy rustdoc_json...
Running eslint on rustdoc JS files
No error code explanation was removed!
tidy: Skipping binary file check, read-only filesystem
tidy error: `/checkout/library/core/src/mem/drop_guard.rs:144` contains `#[test]`; `core` unit tests and benchmarks must be placed into `coretests`
##[error]tidy error: /checkout/library/core/src/mem/drop_guard.rs:133: undocumented unsafe
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.1.1)
linting python files
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:15
  local time: Mon Jul 21 03:40:28 UTC 2025
  network time: Mon, 21 Jul 2025 03:40:28 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

orlp commented:
Some things missing in this PR from my ACP proposal (which makes sense since it was apparently done in parallel):

1. The function type in my ACP is defaulted to `= fn(T)`:
```rust
pub struct WithDrop<T, F: FnOnce(T) = fn(T)> {
```
This is similar to how it is defaulted in [`LazyLock`](https://doc.rust-lang.org/beta/std/sync/struct.LazyLock.html) to make writing static declarations easier.

2. `new` is not marked `const`. There's no reason it shouldn't be.
3. `new` is not marked `#[must_use]`. This is important to prevent accidental instant-dropping.
5. `Clone` is not derived. I think this somewhat makes sense if you think of this type as exclusively a guard, but I think of this [more generally](https://github.com/rust-lang/libs-team/issues/622#issuecomment-3095699609), and under that umbrella there is no reason `Clone` shouldn't be derived.

There are also some differences:
1. `Debug` is not implemented, but rather derived. This gives rather ugly results.
2. I intentionally did not implement `PartialEq, Eq, PartialOrd, Ord, Hash`, but this PR does (through derives). It's not clear to me that these traits should be implemented for this type as it's unclear whether these relationships include the `on_drop` function to compare or not. One user might expect equality to only include the payload, and another might expect it to include the function. As it's written right now it's also dangerous because `Deref` coercion might change equality results (since post-`Deref` only the value is compared, but pre-`Deref` both the value and function).

purplesyringa commented on library/core/src/mem/drop_guard.rs:
```diff
@@ -0,0 +1,166 @@
+    ///
+    /// It is typically preferred to call this function instead of `mem::forget`
+    /// because it will appropriately de-initialize the closure stored in the
+    /// `DropGuard` before returning the stored value.
```
Why draw attention to de-initialization of the closure, but not to de-initialization of the `T` itself? `forget` wouldn't drop `T` either.

yoshuawuyts replied:
I mean: `DropGuard` doesn't de-initialize the `T` either: it returns it. But yeah you're right, that's probably worth clarifying. Do you have a better phrasing to propose here?

purplesyringa commented on library/core/src/mem/drop_guard.rs:
```diff
@@ -0,0 +1,166 @@
+
+// tests copied from https://docs.rs/scopeguard/latest/src/scopeguard/lib.rs.html#1-595
+#[cfg(test)]
+mod tests {
```
Maybe add a test for panics occurring within the closure since `into_inner` takes extra precaution to get it right?

yoshuawuyts replied:
I'm not sure what that test would look like, since I'm not sure what the current panic-while-unwinding behavior is supposed to be. How would you write it?

bjorn3 replied:
`#[test]` inside `library/core` is ignored. Tests for libcore should be in `library/coretests` instead, which can use libstd just fine.

lukaslueg commented on library/core/src/mem/drop_guard.rs:
```diff
@@ -0,0 +1,166 @@
+    #[unstable(feature = "drop_guard", issue = "none")]
+    #[inline]
+    pub fn into_inner(guard: Self) -> T {
+        // Implementation note: this is the same approach that scopeguard takes.
```
personal drive-by comment: future readers of this comment are probably not aware if `scopeguard` refers to a crate, a core-module, or a type somewhere in core/std. IMHO this comment should either be left out entirely or moved to a mod-level comment.

yoshuawuyts commented on library/core/src/mem/drop_guard.rs:
```diff
@@ -0,0 +1,151 @@
+    F: FnOnce(T),
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&**self, f)
```
I don't necessarily love this either - but this is what we seem to be doing for most other wrapper types in the stdlib. I lifted this impl from `Mutex`, but for example the `Pin` type does [the same thing](https://doc.rust-lang.org/src/core/pin.rs.html#1697).

purplesyringa replied:
Please correct me if I'm wrong, but I think the rule of thumb here is "`Debug` is transparent if the wrapper type has no runtime behavior". E.g. types like `Mutex`, `LazyLock`, `ManuallyDrop`, etc. implement `Debug` non-transparently, and `Pin`, `Unique` are transparent. I think `DropGuard` is closer to the former than the latter so we wouldn't break any conventions here.

yoshuawuyts replied:
I'm not sure what the rules here are honestly, but I like where you're going with this. While the the impl for `Mutex` is intransparent, the impl for `MutexGuard` is transparent. I think it's fair to say that `MutexGuard` does have its own runtime behavior. How would that fit into this framework?

yoshuawuyts replied:
Going to mark it as resolved on this PR, but filed it as an open question on the tracking issue: https://github.com/rust-lang/rust/issues/144426

purplesyringa commented on library/core/src/mem/drop_guard.rs:
```diff
@@ -0,0 +1,150 @@
+        // Next we manually read the stored value from the guard.
+        //
+        // SAFETY: this is safe because we've taken ownership of the guard.
+        let value = unsafe { ptr::read(&*guard.inner) };
```
This is probably bikeshedding, but why not `ManuallyDrop::take`? That would get rid of raw pointers.

yoshuawuyts replied:
Yeah no that's a good call! The fewer raw pointers we have, the better. Thank you!

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16423258546/job/46407148715) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46407148715) [(plain)](https://github.com/rust-lang/rust/commit/8fd3010fdf775025015ad1add2a547bc52183458/checks/46407148715/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
....................................................................................... 783/2112
....................................................................................... 870/2112
....................................................................................... 957/2112
....................................... 996/2112
mem::drop_guard_always_drops_value_if_closure_drop_unwinds --- FAILED
.....................................................ii................................ 1084/2112
....................................................................................... 1171/2112
....................................................................................... 1258/2112
....................................................................................... 1345/2112
....................................................................................... 1432/2112
---
....................................................................................... 2041/2112
.......................................................................
failures:

---- mem::drop_guard_always_drops_value_if_closure_drop_unwinds stdout ----
---- mem::drop_guard_always_drops_value_if_closure_drop_unwinds stderr ----

thread 'main' panicked at coretests/tests/mem.rs:830:44:
explicit panic
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic_explicit
   3: coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#1}::panic_cold_explicit
             at /checkout/library/core/src/panic.rs:88:13
   4: coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#1}
             at ./tests/mem.rs:830:44
   5: <core::mem::drop_guard::DropGuard<(), coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#1}> as core::ops::drop::Drop>::drop
             at /checkout/library/core/src/mem/drop_guard.rs:137:9
   6: core::ptr::drop_in_place::<core::mem::drop_guard::DropGuard<(), coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#1}>>
             at /checkout/library/core/src/ptr/mod.rs:804:1
   7: core::ptr::drop_in_place::<coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#2}>
             at /checkout/library/core/src/ptr/mod.rs:804:1
   8: <core::mem::drop_guard::DropGuard<core::mem::drop_guard::DropGuard<(), coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#0}>, coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#2}>>::into_inner
             at /checkout/library/core/src/mem/manually_drop.rs:256:18
   9: coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#3}
             at ./tests/mem.rs:839:9
  10: <core::panic::unwind_safe::AssertUnwindSafe<coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#3}> as core::ops::function::FnOnce<()>>::call_once
             at /checkout/library/core/src/panic/unwind_safe.rs:272:9
  11: std::panicking::catch_unwind::do_call::<core::panic::unwind_safe::AssertUnwindSafe<coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#3}>, ()>
             at /checkout/library/std/src/panicking.rs:589:40
  12: std::panic::catch_unwind::<core::panic::unwind_safe::AssertUnwindSafe<coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#3}>, ()>
             at /checkout/library/std/src/panicking.rs:552:19
  13: coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds
             at ./tests/mem.rs:837:13
  14: coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#0}
             at ./tests/mem.rs:824:59
  15: <coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#0} as core::ops::function::FnOnce<()>>::call_once
             at /checkout/library/core/src/ops/function.rs:253:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    mem::drop_guard_always_drops_value_if_closure_drop_unwinds

test result: FAILED. 2107 passed; 1 failed; 4 ignored; 0 measured; 0 filtered out; finished in 2.69s

error: test failed, to rerun pass `-p coretests --test coretests`
env -u RUSTC_WRAPPER CARGO_ENCODED_RUSTDOCFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Dwarnings\u{1f}-Wrustdoc::invalid_codeblock_attributes\u{1f}--crate-version\u{1f}1.90.0-nightly\t(a68933411\t2025-07-21)" CARGO_ENCODED_RUSTFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Zmacro-backtrace\u{1f}-Csplit-debuginfo=off\u{1f}-Clink-arg=-L/usr/lib/llvm-19/lib\u{1f}-Cllvm-args=-import-instr-limit=10\u{1f}-Clink-args=-Wl,-z,origin\u{1f}-Clink-args=-Wl,-rpath,$ORIGIN/../lib\u{1f}-Alinker-messages\u{1f}--cap-lints=allow\u{1f}--cfg\u{1f}randomized_layouts" RUSTC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustc-clif" RUSTDOC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustdoc-clif" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "test" "--manifest-path" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests/Cargo.toml" "--target-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests_target" "--locked" "--target" "aarch64-unknown-linux-gnu" "-p" "coretests" "-p" "alloctests" "--tests" "--" "-q" exited with status ExitStatus(unix_wait_status(25856))
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:14:01
  local time: Mon Jul 21 17:21:52 UTC 2025
  network time: Mon, 21 Jul 2025 17:21:52 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

yoshuawuyts commented on library/coretests/tests/mem.rs:
```diff
@@ -795,3 +796,47 @@ fn const_maybe_uninit_zeroed() {
+
+    // This will run the closure, which will panic when dropped. This should
+    // run the destructor of the value we passed, which we validate.
+    let _ = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
```
If I'm understanding [this failure](https://github.com/rust-lang/rust/actions/runs/16423258546/job/46407148715?pr=144236) correctly, this line isn't actually catching the unwind the way it should.

Or maybe the test runner in CI mode is getting confused that we're unwinding and catching it. I'm not actually sure what is happening?

yoshuawuyts replied:
I tried recreating this locally using `./x.py test library/coretests`, and I can't seem to. All tests are passing, and I'm not quite sure what's different.

Re-running all tests to see if that makes a difference.

orlp replied:
I can't reproduce locally (not as part of coretest, standalone). The panic still gets printed as we haven't installed a `panic_handler` but it should get caught.

purplesyringa replied:
Nope, you're just testing on cranelift with `panic = "abort"`. Add `#[cfg(panic = "unwind")]` to the test.

yoshuawuyts replied:
I don't often contribute directly to `rust-lang/rust`, so I appreciate the help!

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16429061249/job/46426498483) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46426498483) [(plain)](https://github.com/rust-lang/rust/commit/d8afe5622c221f03c75f28756a85df95c52d162d/checks/46426498483/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
....................................................................................... 783/2112
....................................................................................... 870/2112
....................................................................................... 957/2112
................................... 992/2112
mem::drop_guard_always_drops_value_if_closure_drop_unwinds --- FAILED
.........................................................ii............................ 1080/2112
....................................................................................... 1167/2112
....................................................................................... 1254/2112
....................................................................................... 1341/2112
....................................................................................... 1428/2112
---
....................................................................................... 2037/2112
...........................................................................
failures:

---- mem::drop_guard_always_drops_value_if_closure_drop_unwinds stdout ----
---- mem::drop_guard_always_drops_value_if_closure_drop_unwinds stderr ----

thread 'main' panicked at coretests/tests/mem.rs:830:44:
explicit panic
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic_explicit
   3: coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#1}::panic_cold_explicit
             at /checkout/library/core/src/panic.rs:88:13
   4: coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#1}
             at ./tests/mem.rs:830:44
   5: <core::mem::drop_guard::DropGuard<(), coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#1}> as core::ops::drop::Drop>::drop
             at /checkout/library/core/src/mem/drop_guard.rs:136:9
   6: core::ptr::drop_in_place::<core::mem::drop_guard::DropGuard<(), coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#1}>>
             at /checkout/library/core/src/ptr/mod.rs:804:1
   7: core::ptr::drop_in_place::<coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#2}>
             at /checkout/library/core/src/ptr/mod.rs:804:1
   8: <core::mem::drop_guard::DropGuard<core::mem::drop_guard::DropGuard<(), coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#0}>, coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#2}>>::into_inner
             at /checkout/library/core/src/mem/manually_drop.rs:256:18
   9: coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#3}
             at ./tests/mem.rs:839:9
  10: <core::panic::unwind_safe::AssertUnwindSafe<coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#3}> as core::ops::function::FnOnce<()>>::call_once
             at /checkout/library/core/src/panic/unwind_safe.rs:272:9
  11: std::panicking::catch_unwind::do_call::<core::panic::unwind_safe::AssertUnwindSafe<coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#3}>, ()>
             at /checkout/library/std/src/panicking.rs:589:40
  12: std::panic::catch_unwind::<core::panic::unwind_safe::AssertUnwindSafe<coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#3}>, ()>
             at /checkout/library/std/src/panicking.rs:552:19
  13: coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds
             at ./tests/mem.rs:837:13
  14: coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#0}
             at ./tests/mem.rs:824:59
  15: <coretests::mem::drop_guard_always_drops_value_if_closure_drop_unwinds::{closure#0} as core::ops::function::FnOnce<()>>::call_once
             at /checkout/library/core/src/ops/function.rs:253:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    mem::drop_guard_always_drops_value_if_closure_drop_unwinds

test result: FAILED. 2107 passed; 1 failed; 4 ignored; 0 measured; 0 filtered out; finished in 2.75s

error: test failed, to rerun pass `-p coretests --test coretests`
env -u RUSTC_WRAPPER CARGO_ENCODED_RUSTDOCFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Dwarnings\u{1f}-Wrustdoc::invalid_codeblock_attributes\u{1f}--crate-version\u{1f}1.90.0-nightly\t(883fcfb63\t2025-07-21)" CARGO_ENCODED_RUSTFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}--check-cfg=cfg(llvm_enzyme)\u{1f}-Zmacro-backtrace\u{1f}-Csplit-debuginfo=off\u{1f}-Clink-arg=-L/usr/lib/llvm-19/lib\u{1f}-Cllvm-args=-import-instr-limit=10\u{1f}-Clink-args=-Wl,-z,origin\u{1f}-Clink-args=-Wl,-rpath,$ORIGIN/../lib\u{1f}-Alinker-messages\u{1f}--cap-lints=allow\u{1f}--cfg\u{1f}randomized_layouts" RUSTC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustc-clif" RUSTDOC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/dist/rustdoc-clif" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "test" "--manifest-path" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests/Cargo.toml" "--target-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools/cg_clif/build/sysroot_tests_target" "--locked" "--target" "aarch64-unknown-linux-gnu" "-p" "coretests" "-p" "alloctests" "--tests" "--" "-q" exited with status ExitStatus(unix_wait_status(25856))
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:13:57
  local time: Mon Jul 21 22:18:45 UTC 2025
  network time: Mon, 21 Jul 2025 22:18:45 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

PatchMixolydic commented on library/core/src/mem/drop_guard.rs:
```diff
@@ -0,0 +1,149 @@
+
+    /// Consumes the `Guard`, returning the wrapped value.
+    ///
+    /// This will not execute the closure. This is implemented as a static
+    /// method to prevent any potential conflicts with any other methods called
```
Nit:
```suggestion
    /// This will not execute the closure. This is implemented as an associated
    /// function to prevent any potential conflicts with any other methods called
```

yoshuawuyts commented:
Filed a tracking issue in https://github.com/rust-lang/rust/issues/144426, updated the `#[unstable]` attributes, and removed the `= fn() -> T` default parameter from the type. I never added the `Clone` impl to begin with, so I believe with that this PR should be in compliance with the T-Libs-API acceptance criteria defined in https://github.com/rust-lang/libs-team/issues/622.

Once the latest CI run is green, I think this should be good to merge!

Mark-Simulacrum commented:
r=me with commits squashed

Mark-Simulacrum commented:
@bors r+
